src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -403,105 +403,124 @@ exprt string_constraint_generatort::add_axioms_for_correct_number_format(
   axioms.push_back(
     implies_exprt(correct, str.axiom_for_is_shorter_than(max_size)));
 
-  // forall 1 <= qvar < |str| . correct => is_digit_with_radix(str[qvar], radix)
-  symbol_exprt qvar=fresh_univ_index("number_format", index_type);
-  exprt is_digit=is_digit_with_radix(str[qvar], radix);
-  string_constraintt all_digits(
-    qvar, from_integer(1, index_type), str.length(), correct, is_digit);
-  axioms.push_back(all_digits);
+//  // forall 1 <= qvar < |str| . correct => is_digit_with_radix(str[qvar], radix)
+//  symbol_exprt qvar=fresh_univ_index("number_format", index_type);
+//  exprt is_digit=is_digit_with_radix(str[qvar], radix);
+//  string_constraintt all_digits(
+//    qvar, from_integer(1, index_type), str.length(), correct, is_digit);
+//  axioms.push_back(all_digits);
+
+  for(std::size_t index=1; index<max_size; ++index)
+  {
+    const exprt index_expr=from_integer(index, index_type);
+    /// index < length && correct => is_digit(str[index])
+    implies_exprt character_at_index_is_digit(
+      and_exprt(
+        binary_relation_exprt(index_expr, ID_lt, str.length()), correct),
+      is_digit_with_radix(str[index], radix));
+    axioms.push_back(character_at_index_is_digit);
+  }
 
   return correct;
 }
 
 /// add axioms corresponding to the Integer.parseInt java function
 /// \param f: a function application with either one string expression or one
-///   string expression and an expression for the radix
+///   string expression and an integer expression for the radix
 /// \return an integer expression
 exprt string_constraint_generatort::add_axioms_for_parse_int(
   const function_application_exprt &f)
 {
   PRECONDITION(f.arguments().size()==1 || f.arguments().size()==2);
-  string_exprt str=get_string_expr(f.arguments()[0]);
+  const string_exprt str=get_string_expr(f.arguments()[0]);
+  const typet &type=f.type();
+  PRECONDITION(type.id()==ID_signedbv);
```

Comment:
I wouldn't relax the condition here. The main reason being that there is no reason to do it from a Java point of view (as all integers are signed anyway). Also, I'm not sure what it would mean to have an unsigned return value when we are parsing a negative integer.